### PR TITLE
Expression tree

### LINKED ISSUE #1132: expression tree in c++ and python

### PR DESCRIPTION
**Pull request by raghav-dalmia**
Fixes #1132 

#### Checklist:
- [X] 4 space indentation.
- [X] Coding conventions are followed.
- [X] Input is taken dynamically.
- [X] Sample Input / Output is added at the end of the file.
- [X] Logic Documentation (Comments).
- [X] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
-  add Expression tree code

#### Languages Used:
-  c++
-  python

#### Files Added:
-  Expression_Tree.cpp
-  Expression_Tree.py

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

**Comment by raghav-dalmia** [comment]
@jainaman224, I have closed the previous PR. Please review this one.

File: Expression_Tree/Expression_Tree.cpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+//expression tree class
+class exp_tree
+{
+    string postfix_exp;
+    node *root;
+    public:
+    exp_tree(string postfix_exp)
+    {
+        this->postfix_exp = postfix_exp;
```
Comment (marwahmanbir) [inline Expression_Tree/Expression_Tree.cpp]: Add space around ``->``.

File: Expression_Tree/Expression_Tree.cpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            return true; // then return true
+        }
+        return false; // else return false
+    }
+
+    void create_tree(string exp)
+    {
+        int len = exp.length();
+        stack<node *> s; // store those operator node whose any child node is NULL
+        root = new node(exp[len-1]);
```
Comment (marwahmanbir) [inline Expression_Tree/Expression_Tree.cpp]: Space around ``-``.

File: Expression_Tree/Expression_Tree.cpp [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        this->postfix_exp = postfix_exp;
+        root = NULL;
+        create_tree(postfix_exp);
+    }
+
+    private:
+    bool is_operator(char c)
+    {
+        if(c == '+' || c == '-' || c == '*' || c == '/' || c == '^') // if given char is operator
```
Comment (marwahmanbir) [inline Expression_Tree/Expression_Tree.cpp]: Space after ``if`` at all the places.

File: Expression_Tree/Expression_Tree.cpp [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        return false; // else return false
+    }
+
+    void create_tree(string exp)
+    {
+        int len = exp.length();
+        stack<node *> s; // store those operator node whose any child node is NULL
+        root = new node(exp[len-1]);
+        s.push(root); // last character of postfix expression is always an operator
+        for(int i = len-2;i >= 0;i--) // travel on rest of the postfix expression
```
Comment (marwahmanbir) [inline Expression_Tree/Expression_Tree.cpp]: Space after operators and semicolons. Also after ``for`` at all the places.

File: Expression_Tree/Expression_Tree.cpp [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        {
+            node* curr_node = s.top();
+            if(curr_node->right == NULL) // if right node of current node is NULL
```
Comment (marwahmanbir) [inline Expression_Tree/Expression_Tree.cpp]: Space after ``if`` and around ``->``.

File: Expression_Tree/Expression_Tree.cpp [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+            {
+                node *temp = new node(exp[i]);
+                curr_node->right = temp;
```
Comment (marwahmanbir) [inline Expression_Tree/Expression_Tree.cpp]: Fix spacing.

File: Expression_Tree/Expression_Tree.cpp [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+                if(is_operator(exp[i]))
```
Comment (marwahmanbir) [inline Expression_Tree/Expression_Tree.cpp]: fix

File: Expression_Tree/Expression_Tree.cpp [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+                {
+                    s.push(temp);
+                }
+            }
+            else // if left node of current node is NULL
+            {
+                node *temp = new node(exp[i]);
+                curr_node->left = temp;
```
Comment (marwahmanbir) [inline Expression_Tree/Expression_Tree.cpp]: fix

File: Expression_Tree/Expression_Tree.cpp [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+                // if no child node of current node is NULL
+                s.pop(); // pop current from stack
+                if(is_operator(exp[i]))
```
Comment (marwahmanbir) [inline Expression_Tree/Expression_Tree.cpp]: fix

File: Expression_Tree/Expression_Tree.cpp [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+                {
+                    s.push(temp);
+                }
+            }
+        }
+    }
+    void inorder_traversal(node *head) // inorder traversal of expression tree
+    {
+        // inorder traversal => left,root,right
+        if(head->left != NULL)
```
Comment (marwahmanbir) [inline Expression_Tree/Expression_Tree.cpp]: Fix

File: Expression_Tree/Expression_Tree.cpp [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+        {
+            inorder_traversal(head->left);
```
Comment (marwahmanbir) [inline Expression_Tree/Expression_Tree.cpp]: Fix

File: Expression_Tree/Expression_Tree.cpp [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+        }
+        cout << head->data << " ";
```
Comment (marwahmanbir) [inline Expression_Tree/Expression_Tree.cpp]: Fix

File: Expression_Tree/Expression_Tree.cpp [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+        if(head->right != NULL)
```
Comment (marwahmanbir) [inline Expression_Tree/Expression_Tree.cpp]: Fix

File: Expression_Tree/Expression_Tree.cpp [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+        {
+            inorder_traversal(head->right);
```
Comment (marwahmanbir) [inline Expression_Tree/Expression_Tree.cpp]: Here also

File: Expression_Tree/Expression_Tree.py [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+            return self.arr[-1]
+        except:
+            pass
+
+    def size(self):
+        return len(self.arr)
+
+# node class for expression tree
+
+
```
Comment (marwahmanbir) [inline Expression_Tree/Expression_Tree.py]: Remove the two extra blank lines.

File: Expression_Tree/Expression_Tree.py [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+class node:
+    def __init__(self, data):
+        self.data = data
+        self.left = None
+        self.right = None
+
+# expression tree class
+
+
```
Comment (marwahmanbir) [inline Expression_Tree/Expression_Tree.py]: Remove the two extra blank lines.

**Comment by raghav-dalmia** [comment]
@marwahmanbir, can you please specify the code style that you want me to follow? I use autopep8 style for python.

**Comment by raghav-dalmia** [comment]
@jainaman224 I have a suggestion, you should mention the exact code style in README.md so as to one style over complete repo.

**Comment by raghav-dalmia** [comment]
I will surely make these changes when I get time. Thank you.

**Comment by jainaman224** [comment]
@raghav-dalmia Will try to be more detailed but due to too much work load. I am unable to give much time to each PR.

**Comment by raghav-dalmia** [comment]
done @marwahmanbir @jainaman224.

**Comment by MastersAbh** [comment]
@marwahmanbir have a look.

**Pull request merged**